gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemBase.java

```diff
@@ -1483,6 +1489,41 @@ private Credential getCredential(Configuration config)
       }
     }
 
+    // If impersonation service account exists, then use current credential to request access token
+    // for the impersonating service account.
+    String impersonationServiceAccount =
+        HadoopCredentialConfiguration.getImpersonationServiceAccount(config, GCS_CONFIG_PREFIX);
+    if (impersonationServiceAccount != null) {
```

Comment:
We should extract this logic in `util` project, maybe in the `ImpersonatedCredentialsFactory` class?

Reply:
Extracted in CredentialFromAccessTokenProviderClassFactory as a code reusable wrapper for now cuz the impersonate service account essentially utilizes AccessTokenProvider as a credential source. 

Http transport/initializer still need to be extracted here to get away from circular import.